apps/web/components/flow/pane.tsx

```diff
@@ -1,127 +1,61 @@
-
-  /**
-   * redux state
-   */
-  const user = useAppSelector((state) => state.user)
-
-  /**
-   * builtin react flow hooks that handle node/edge movement
-   * here, the variables `nodes` and `edges` store the current state of all
-   * nodes and edges on-screen.
-   */
-  const [nodes, setNodes, onNodesChange] = useNodesState(graph.flowNodes)
-  const [edges, setEdges] = useEdgesState(graph.flowEdges)
-
-  useEffect(() => {
-    setEdges(graph.flowEdges)
-    const newNodes = redrawGraph({
-      nodes: graph.flowNodes,
-      edges: graph.flowEdges,
-    }).nodes
-
-    if (user.id && graph.id) {
-      // updates CSS after
-      // Redux state user is not detecting changes, so rely on API call for
-      // updated user.
-      Promise.all([
-        api.user.getById(user.id),
-        api.graph.canTakeModules(graph.id),
-      ]).then(([user, canTake]) =>
-        setNodes(getCSS(newNodes, user.modulesDone, user.modulesDoing, canTake))
-      )
-    }
-  }, [graph.flowNodes, graph.id])
+  /** TODO: use node types to set state */
+  const nodeTypes = useMemo(() => ({ moduleNode: ModuleNode }), [])
 
-  // Update CSS of nodes, if modulesDone/modulesDoing has changed
-  // Uses redux state user.
-  useEffect(() => {
-    const done = user.modulesDone
-    const doing = user.modulesDoing
-    if (graph.id) {
-      trpc
-        .query('graph/can-take-modules', graph.id)
-        .then((canTake) => setNodes(getCSS(nodes, done, doing, canTake)))
-    }
-  }, [user.modulesDone, user.modulesDoing])
-
-  /**
-   * Fit view for ANY graph change
-   * - nodes
-   * - edges
-   * - change graph
-   */
+  /** hooks */
   const reactFlow = useReactFlow()
-  useEffect(() => {
-    reactFlow.fitView({ maxZoom: 1 })
-  }, [nodes, edges])
+  const dispatch = useAppDispatch()
+  const { flowNodes, flowEdges, selectedCodes } = useAppSelector(
+    (s) => s.modtree.graph
+  )
 
-  /**
-   * called when user drops a module node. (after having dragged it)
-   */
+  /** Fit view for ANY graph change */
+  useEffect(() => reactFlow.fitView({ maxZoom: 1 }), [flowNodes, flowEdges])
+
+  /** called when user drops a module node. (after having dragged it) */
   const onNodeDragStop = (
     _event: MouseEvent,
     _droppedNode: Node,
     droppedNodes: GraphFlowNode[]
   ) => {
-    setNodes(
-      redrawGraph({
-        nodes,
-        edges,
-      }).nodes.map((node) => ({
-        ...node,
-        selected: droppedNodes.map((d) => d.id).includes(node.id),
-      }))
-    )
+    const droppedIds = droppedNodes.map((d) => d.id)
+    // revert selection state to before dragging
+    const reselect = (node: GraphFlowNode) => ({
+      ...node,
+      selected: droppedIds.includes(node.id),
+    })
+    // update redux state
+    dispatch(setNodesAndEdges(flowNodes.map(reselect)))
   }
 
   return (
     <ReactFlow
-      nodes={nodes}
-      edges={edges}
+      nodes={flowNodes}
+      edges={flowEdges}
```

Comment:
directly integrate redux with react-flow.
no more retarded useEffect calls just to sync redux state with react flow state.